Replaces scripting global protection with call sandboxing

weswigham opened the pull request:
As mentioned may be worth doing in #1725.

This does impose a larger performance penalty on scripting calls, though - a new environment is copied for each call, and that's fairly expensive. Are there any thoughts on this approach?

badboy commented:
Do you have benchmarks how much this impacts performance? I don't think we should do everything we can just to stop the user from shooting itself in the foot. Protecting from globals was an easy fix (and a common mistake) so that was fine.

weswigham commented:
At this point it's not terrible, it's a static overhead of 0.086 milliseconds per invocation of eval on my VM. For a noop script, that's still double to triple the invocation time, though. I may be able to bring it down some by moving the deepcopy implementation into C.

weswigham commented:
New implementation using a lazy copy has almost 0 overhead compared to previous versions, while accomplishing the duty of sandboxing admirably. Overhead is, instead of being forward loaded as before, moved to when accessing globals/global tables (just as before, via metamethods). Lua overhead per call should now be very similar to prior to sandboxing (I've added the overhead of 3 lua function calls to each EVAL (`sandbox` -> `setfenv` -> `setmetatable`), which should be negligible).

I need to further enhance the tests to ensure the security of this approach, though. I'm less confident in how strong the separation between copies is.
